app/models/user.rb

```diff
@@ -506,39 +520,77 @@ def membership_packet_sent?
-  # @return [Boolean] - result of updating :date_membership_packet_sent
+  # @param date_sent [Time] when the packet was sent. default = now
+  # @return [true,false] result of updating :date_membership_packet_sent
   def toggle_membership_packet_status(date_sent = Time.zone.now)
     new_sent_time = membership_packet_sent? ? nil : date_sent
     update(date_membership_packet_sent: new_sent_time)
   end
 
 
-  # TODO this doesn't belong in User.  but not sure yet where it does belong.
-  # FIXME - what if someone is in the grace period? is logic correct?
-  def file_uploaded_during_this_membership_term?
-    return false unless current_member? || in_grace_period?
+  # Was a file uploaded during the right time frame (which is based on the membership_status)?
+  # If the user is a current_member, time frame is the current membership term (includes the first and last days)
+  # If the user is not (yet) a member, time frame is any time including today
+  # If the user is a member in the grace period, time frame is after the last day of the most recent time frame
+  # If the user is a former member, time frame is after the last day of the most recent membership
+  # Else is false.
+  # @todo this doesn't belong in User.  but not sure yet where it does belong.
+  # @todo Since there is more and more logic (more business rules) that varies based on the membership_status,
+  #   it may be time to create subclasses/mixins (modules) for each membership_status
+  #
+  # @return [true,false]
+  def file_uploaded_during_right_time?
+    return false if uploaded_files.blank?
 
-    if current_member?
-      file_uploaded_on_or_after?(current_membership.first_day)
-    else
-      # is in_grace_period
-      file_uploaded_on_or_after?(most_recent_membership.first_day)  # FIXME is this correct?
+    case membership_status
+      when STATE_CURRENT_MEMBER
+        file_uploaded_during_this_membership_term?
+      when STATE_IN_GRACE_PERIOD, STATE_FORMER_MEMBER
+        file_uploaded_on_or_after?(membership_last_day + 1.day)
+      when STATE_NOT_A_MEMBER
+        !uploaded_files.blank?
+      else
+        false
     end
   end
 
+  # @todo this doesn't belong in User.  but not sure yet where it does belong.
+  # @return [true,false]
+  def file_uploaded_during_this_membership_term?
+    return false if uploaded_files.blank? || !current_member?
+
+    file_uploaded_on_or_after?(current_membership.first_day, end_date: current_membership.last_day)
+  end
 
-  def file_uploaded_on_or_after?(the_date = Date.current)
+
+  # Was a file uploaded by the user on or after the given date AND on or before the given end date?
+  # If no end date is given, the default end date is today (Date.current)
+  # true iff the_date <= date any file was uploaded <= end_date
+  #
+  # @param the_date [Date]
+  # @param end_date [Date] optional) default value is Date.current
+  # @return [true,false]
+  def file_uploaded_on_or_after?(the_date = Date.current, end_date: Date.current)
```

Comment:
`end_date` is optional, and will likely be omitted most of the time, so "..within_date_range" isn't quite right.  I agree that "...on_or_after?" isn't great either.  I'll put a @todo in so that it can be renamed later.